lib/src/network/header_proxy_protocol.rs

```diff
@@ -0,0 +1,166 @@
+use std::net::{SocketAddr, SocketAddrV4, SocketAddrV6};
+use std::fmt;
+
+pub enum ProxyProtocolHeader {
+  V1(HeaderV1),
+  V2(HeaderV2),
+}
+
+impl ProxyProtocolHeader {
+  // Use this method to writte the header in the backend socket
+  pub fn into_bytes(&self) -> Vec<u8> {
+    match *self {
+      ProxyProtocolHeader::V1(ref header) => header.into_bytes(),
+      ProxyProtocolHeader::V2(_) => unimplemented!(),
+    }
+  }
+}
+
+/// Indicate the proxied INET protocol and family
+#[derive(Debug, PartialEq, Eq)]
+pub enum ProtocolSupportedV1 {
+  TCP4,    // TCP over IPv4
+  TCP6,    // TCP over IPv6
+  UNKNOWN, // unsupported,or unknown protocols
+}
+
+impl fmt::Display for ProtocolSupportedV1 {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    match *self {
+      ProtocolSupportedV1::TCP4 => write!(f, "TCP4"),
+      ProtocolSupportedV1::TCP6 => write!(f, "TCP6"),
+      ProtocolSupportedV1::UNKNOWN => write!(f, "UNKNOWN"),
+    }
+  }
+}
+
+/// Proxy Protocol header for version 1 (text version)
+/// Example:
+/// - TCP/IPv4: `PROXY TCP4 255.255.255.255 255.255.255.255 65535 65535\r\n`
+/// - TCP/IPv6: `PROXY TCP6 ffff:f...f:ffff ffff:f...f:ffff 65535 65535\r\n`
+/// - Unknown: `PROXY UNKNOWN\r\n`
+#[derive(Debug)]
+pub struct HeaderV1 {
+  protocol: ProtocolSupportedV1,
+  addr_src: SocketAddr,
+  addr_dst: SocketAddr,
+}
+
+const PROXY_PROTO_IDENTIFIER: &'static str = "PROXY";
+
+impl HeaderV1 {
+
+  pub fn new (addr_src: SocketAddr, addr_dst: SocketAddr) -> Self {
+    let protocol = if addr_dst.is_ipv6() {
+      ProtocolSupportedV1::TCP6
+    } else if addr_dst.is_ipv4() {
+      ProtocolSupportedV1::TCP4
+    } else {
+      ProtocolSupportedV1::UNKNOWN
+    };
+
+    HeaderV1 {
+      protocol,
+      addr_src,
+      addr_dst,
+    }
+  }
+
+  pub fn into_bytes(&self) -> Vec<u8> {
+    if self.protocol.eq(&ProtocolSupportedV1::UNKNOWN) {
+      format!("{} {}\r\n",
+          PROXY_PROTO_IDENTIFIER,
+          self.protocol,
+      ).into_bytes()
+    } else {
+      format!("{} {} {} {} {} {}\r\n",
+        PROXY_PROTO_IDENTIFIER,
+        self.protocol,
+        self.addr_src.ip(),
+        self.addr_dst.ip(),
+        self.addr_src.port(),
+        self.addr_dst.port(),
+      ).into_bytes()
+    }
+  }
+}
+
+#[cfg(test)]
+mod test {
+
+  use super::*;
+  use std::net::{IpAddr, Ipv4Addr, SocketAddr, Ipv6Addr};
+
+  #[test]
+  fn it_should_return_a_correct_header_with_ipv4() {
+    let header = HeaderV1::new(
+      SocketAddr::new(IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1)), 80),
```

Comment:
maybe use two different IP addresses for source and destination in the tests?

Reply:
Ok

Reply:
Done